apps/yggdrasil/lib/yggdrasil/user.ex

```diff
@@ -29,14 +36,105 @@ defmodule Yggdrasil.User do
   end
 
   @doc """
-  adds password at end of chain protects the hashpwsalt from seeing a nil value
-  in the case of missing password field.
+  Creates the user and assigns the default roles of "player"
+  """
+  def create_with_default_role(attributes) do
+    changeset = User.create_changeset(%User{}, attributes)
+
+    with {:ok, user} <- Repo.insert(changeset),
+          :ok        <- assign_role(user, "player"),
+     do: {:ok, user}
+  end
+
+  def assign_role(user, role_name) do
+    role = Repo.get_by!(Role, name: role_name)
+
+    user_role = UserRole.changeset(%UserRole{},
+                                   %{user_id: user.id, role_id: role.id})
+    case Repo.insert(user_role) do
+      {:ok, _role} -> :ok
+      error        -> error
+    end
+  end
+
+  def load_permissions(user) do
+    query = from ur in UserRole,
+            where: ur.user_id ==  ^user.id,
+            preload: [role: [:role_permissions]]
+
+    permissions = query
+    |> Repo.all()
+    |> Enum.flat_map(fn ur -> ur.role.role_permissions end)
+    |> Enum.group_by(fn rp -> rp.resource end)
+    |> Enum.map(fn {r, rps} -> {String.to_atom(r), Enum.map(rps, fn rp -> String.to_atom(rp.permission) end)} end)
+    |> Enum.into(%{})
+
+    %{ user | permissions: permissions}
+  end
+
+  @doc ~S"""
+  Checks the list of resource/permissions pair given against what the user
+  has been granted and returns true if and only if all permissions supplied
+  have been granted for the user.
+
+  ## roles
+  A user can have one or more roles with each role having a set of resources
+  and a set of resources and those resources having a set of permissions. Roles
+  are not checked directly and only serve as a grouping mechanism for resources.
+
+  Since the user can have multiple roles, the permission set checked for a given
+  resource is the unique set across all the roles for that resource.
+
+  So if the user has `Role1` and `Role2` defined as follows:
+
+  * Role1
+    * foo
+      * :read
+      * :write
+  * Role2
+    * foo
+      * :read
+      * :all
+
+  The set of permissions checked for foo will be `[:read, :write, :all]`
+
+  ## Examples
+      # checking a single resource
+      User.is_granted? user, foo: [:read]
+
+      # checking multiple resources is allowed as well
+      User.is_granted? user, foo: [:read], bar: [:read, :write]
   """
-  def hash_password(changeset = %{:valid? => false}) do
+  def is_granted?(user, res_perms) do
+    results = Enum.map res_perms, fn {res, perms} ->
+      res_perm_set = user.permissions
+      |> Map.get(res)
+      |> MapSet.new
+
+      # turn perm atoms into strings
+      perm_set = MapSet.new(perms)
+
+      cond do
+        MapSet.size(perm_set) == 0 ->
+          raise "permission set for resource #{res} can't be empty"
+        MapSet.size(res_perm_set) == 0 ->
+          raise "supplied permission set to check for resource #{res} can't be empty"
```

Comment:
I believe we want raise exceptions in these two cases. It's the two edge cases where we would want to catch them quickly and it be loud about it.

There is an edge case here that I need to test out. I'm not exactly sure what happens if the `res` passed is not in the map. I believe that will return nil, which will then produce a set of `[nil]` - in which case everything will work but return false which it should do I think. Which happy accident that needs to be more explicit, so this will change again. However if it produces an empty set that will raise an error and only in that specific case I think it should not error. So I think that pipeline above needs to be broken up.

However if a key does exist for the resource but the permissions list is empty/nil (which I didn't check for nil) that is exceptional case that should not happen and I think it would be better to error than simply ignore it and treat is as false.

Same with passing in an empty list of permissions when using the function, originally I returned false, but I don't think there is a valid reason for doing this and chose to enforce passing in something. It seems right but maybe it's not.